src/rdkafka_idempotence.c

```diff
@@ -261,18 +277,45 @@ void rd_kafka_idemp_pid_fsm (rd_kafka_t *rk) {
                         return;
                 }
 
-                rd_rkb_dbg(rkb, EOS, "GETPID", "Acquiring ProducerId");
-
-                err = rd_kafka_InitProducerIdRequest(
-                        rkb,
-                        rk->rk_conf.eos.transactional_id,
-                        rd_kafka_is_transactional(rk) ?
-                        rk->rk_conf.eos.transaction_timeout_ms : -1,
-                        rd_kafka_pid_valid(rk->rk_eos.pid) ?
-                        &rk->rk_eos.pid : NULL,
-                        errstr, sizeof(errstr),
-                        RD_KAFKA_REPLYQ(rk->rk_ops, 0),
-                        rd_kafka_handle_InitProducerId, NULL);
+                if (rd_kafka_is_transactional(rk)) {
+                        /* If this is a transactional producer and the
+                         * PID-epoch needs to be bumped we'll require KIP-360
+                         * support on the broker, else raise a fatal error. */
+
+                        if (rd_kafka_pid_valid(rk->rk_eos.pid))
+                                rd_rkb_dbg(rkb, EOS, "GETPID",
+                                           "Re-acquiring ProducerId bump "
+                                           "for %s",
+                                           rd_kafka_pid2str(rk->rk_eos.pid));
+                        else
+                                rd_rkb_dbg(rkb, EOS, "GETPID",
+                                           "Acquiring ProducerId");
+
+                        err = rd_kafka_InitProducerIdRequest(
+                                rkb,
+                                rk->rk_conf.eos.transactional_id,
+                                rk->rk_conf.eos.transaction_timeout_ms,
+                                rd_kafka_pid_valid(rk->rk_eos.pid) ?
+                                &rk->rk_eos.pid : NULL,
+                                errstr, sizeof(errstr),
+                                RD_KAFKA_REPLYQ(rk->rk_ops, 0),
+                                rd_kafka_handle_InitProducerId, NULL);
+
+                        if (err == RD_KAFKA_RESP_ERR__UNSUPPORTED_FEATURE &&
+                            rd_kafka_pid_valid(rk->rk_eos.pid))
+                                is_fatal = rd_true;
+                } else {
+                        rd_rkb_dbg(rkb, EOS, "GETPID", "Acquiring ProducerId");
+
+                        err = rd_kafka_InitProducerIdRequest(
+                                rkb,
+                                rk->rk_conf.eos.transactional_id,
```

Comment:
if this isn't a transactional producer, what is the transactional_id being passed for? i assume this is NULL or something in this case, so correct, but maybe write that explicitly.